separate tutorials into each document and add images (#1230)
vdaas/vald: docs/tutorial/vald-agent-standalone-on-k8s.md

# Vald Agent Standalone on Kubernetes

This article will show you how to deploy a standalone Vald Agent using Helm and run it on your kubernetes cluster.

## Overview

> Vald is made up of multiple micro services.
> In the [Get Started](../tutorial/get-started.md), you may use 4 kinds of microsevices to deploy Vald.
> In this case, you use only 1 micro service, `Vald Agent` that is the core component for Vald named as `vald-agent-ngt`, to deploy.
> The below image shows the architecture image of this case.

Reviewer: ```suggestion
Vald is made up of multiple microservices.
In the [Get Started](../tutorial/get-started.md), you may use 4 kinds of microsevices to deploy Vald.
In this case, you use only 1 microservice, `Vald Agent` that is the core component for Vald named as `vald-agent-ngt`, to deploy.
The below image shows the architecture image of this case.
```

<img src="../../assets/docs/tutorial/vald-agent-standalone-on-k8s.png">

> The 5 steps to Vald Agent Standalone on Kubernetes with Vald:
> 1. [Check and satisfy the requirements](#Requirements)
> 1. [Prepare kubernetes cluster](#Prepare-the-kubernetes-cluster)
> 1. [Deploy Vald Agent Standalone on kubernetes cluster](#Deploy-Vald-Agent-Standalone-on-Kubernetes-Cluster)
> 1. [Running exapmle code](#Running-Exapmle-Code)
> 1. [Cleanup](#Cleanup)

Reviewer: ```suggestion
The 5 steps to Vald Agent Standalone on Kubernetes with Vald:
1. [Check and satisfy the requirements](#Requirements)
1. [Prepare kubernetes cluster](#Prepare-the-kubernetes-cluster)
1. [Deploy Vald Agent Standalone on kubernetes cluster](#Deploy-Vald-Agent-Standalone-on-Kubernetes-Cluster)
1. [Running example code](#Running-Example-Code)
1. [Cleanup](#Cleanup)
```

## Requirements

- kubernetes: v1.19 ~
- go: v1.15 ~
- helm: v3 ~
- libhdf5 (_only required for get started_)

> Helm is used to deploying Vald on your kubernetes and Hdf5 is used to decode the sample data file to run the example.<br>

Reviewer: ```suggestion
Helm is used to deploying Vald on your Kubernetes and Hdf5 is used to decode the sample data file to run the example.<br>
```
same 🙏 
https://github.com/vdaas/vald/pull/1230/files#r633278044

If helm or hdf5 is not installed, please install [helm](https://helm.sh/docs/intro/install) and [hdf5](https://www.hdfgroup.org/).

<details><summary>Installation command for helm</summary><br>

```bash
curl https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 | bash
```

</details>

<details><summary>Installation command for hdf5</summary><br>

```bash
# yum
yum install -y hdf5-devel

# apt
apt-get install libhdf5-serial-dev

# homebrew
brew install hdf5
```

</details>

## Prepare the Kubernetes Cluster

1. Prepare kubernetes cluster

>     To complete get started, the Kubernetes cluster is required.<br>
>     Vald will run on Cloud Service such as GKE, AWS.
>     In the sense of trying to "Get-Started", [k3d](https://k3d.io/) or [kind](https://kind.sigs.k8s.io/) are easy kubernetes tools to use.
> 
> ## Deploy Vald Agent Standalone on Kubernetes Cluster
> 
> This chapter will show you how to deploy a standalone Vald Agent using Helm and run it on your kubernetes cluster. <br>
> This chapter uses [NGT](https://github.com/yahoojapan/ngt) as Vald Agent to perform vector insertion operation, indexing and searching operations.<br>

Reviewer: ```suggestion
    To complete get started, the Kubernetes cluster is required.<br>
    Vald will run on Cloud Service such as GKE, AWS.
    In the sense of trying to "Get-Started", [k3d](https://k3d.io/) or [kind](https://kind.sigs.k8s.io/) are easy Kubernetes tools to use.

## Deploy Vald Agent Standalone on Kubernetes Cluster

This chapter will show you how to deploy a standalone Vald Agent using Helm and run it on your Kubernetes cluster. <br>
This chapter uses [NGT](https://github.com/yahoojapan/ngt) as Vald Agent to perform vector insertion operation, indexing and searching operations.<br>
```

same 🙏 
https://github.com/vdaas/vald/pull/1230/files#r633278044

1. Clone the repository

>     To use the `deployment yaml` for deploy, let's clone [`vdaas/vald`](https://github.com/vdaas/vald.git) repository.

Reviewer: ```suggestion
    To use the `deployment yaml` to deploy, let's clone [`vdaas/vald`](https://github.com/vdaas/vald.git) repository.
```

    ```bash
    git clone https://github.com/vdaas/vald.git
    cd vald
    ```

1. Deploy Vald Agent Standalonw using Helm

    There is the [values.yaml](https://github.com/vdaas/vald/blob/master/example/helm/values-standalone-agent-ngt.yaml) to deploy standalone Vald Agent.
    Each component can be disabled by setting the value `false` to the `[component].enabled` field.
    This is useful for deploying standalone Vald Agent NGT pods.

    ```bash
    helm repo add vald https://vald.vdaas.org/charts
    helm install vald-agent-ngt vald/vald --values example/helm/values-standalone-agent-ngt.yaml
    ```

1. Verify

    ```bash
    kubectl get pods
    ```

    <details><summary>Example output</summary><br>
    If the deployment is successful, Vald Agent component should be running.

    ```bash
    NAME               READY   STATUS    RESTARTS   AGE
    vald-agent-ngt-0   1/1     Running   0          20m
    vald-agent-ngt-1   1/1     Running   0          20m
    vald-agent-ngt-2   1/1     Running   0          20m
    vald-agent-ngt-3   1/1     Running   0          20m
    ```

    </details>

## Running Exapmle Code

1. Port Forward

>     At first, port-foward is required to make request from your local environment possible.

Reviewer: ```suggestion
    At first, port-foward is required to make a request from your local environment possible.
```

Reviewer 2: I think `make request` is correct

Reviewer 2: ```suggestion
    At first, port forwarding the vald-lb-gateway is required to make request to the Vald cluster from your local environment.
```

Author: `to the Vald cluster` is not necessary for this sentence 🤔

    ```bash
    kubectl port-forward service/vald-agent-ngt 8081:8081
    ```

1. Download dataset

    Download [fashion-mnist](https://github.com/zalandoresearch/fashion-mnist) that is used as dataset for indexing and search query.

    ```bash
    # move to working directory
    cd example/client/agent

    # download fashion-mnist testing dataset
    wget http://ann-benchmarks.com/fashion-mnist-784-euclidean.hdf5
    ```

1. Run

>     We use [`example/client/agent/main.go`](https://github.com/vdaas/vald/blob/master/example/client/agent/main.go) to run the example.<br>
>     This example will insert and index 400 vectors into the Vald from the fashion-mnist dataset via grpc.
>     And then after waiting for indexing, it will request for searching nearest vector at 10 times.
>     You will get the 10 nearest neighbor vectors for each search query.<br>
>     Run example codes by executing below command.

Reviewer: ```suggestion
    We use [`example/client/agent/main.go`](https://github.com/vdaas/vald/blob/master/example/client/agent/main.go) to run the example.<br>
    This example will insert and index 400 vectors into the Vald from the fashion-mnist dataset via gRPC.
    And then after waiting for indexing, it will request for searching the nearest vector 10 times.
    You will get the 10 nearest neighbor vectors for each search query.<br>
    Run example codes by executing the below command.
```

                    },
                })
                if err != nil {
                    glg.Fatal(err)
                }
            }
            ```

            </details>

        - Wait until indexing finish.
            <details><summary>example code</summary><br>

            ```go
            wt := time.Duration(indexingWaitSeconds) * time.Second
            glg.Infof("Wait %s for indexing to finish", wt)
            time.Sleep(wt)
            ```

            </details>

        - [Optional] Indexing manually instead of waiting for auto indexing
        You can set Agent NGT configuration `auto_index_duration_limit` and `auto_index_check_duration` for auto indexing.
        In this example, you can create index manually using `CreateAndSaveIndex()` mthod in the client library
            <details><summary>example code</summary><br>

            ```go
            _, err = client.CreateAndSaveIndex(ctx, &payload.Control_CreateIndexRequest{
                PoolSize: uint32(insertCount),
            })
            if err != nil {
                glg.Fatal(err)
            }
            ```

            </details>

    1. Search

>         - Search 10 neighbor vectors for each 20 test datasets and return list of neighbor vector.
> 
>         - When getting approximate vectors, the Vald client sends search config and vector to the server via gRPC.
>             <details><summary>example code</summary><br>

Reviewer: ```suggestion
        - Search 10 neighbor vectors for each 20 test datasets and return a list of the neighbor vectors.

        - When getting approximate vectors, the Vald client sends search config and vector to the server via gRPC.
            <details><summary>example code</summary><br>
```

    ```bash
    git clone https://github.com/vdaas/vald.git
    cd vald
    ```

1. Deploy Vald Agent Standalone using Helm

    There is the [values.yaml](https://github.com/vdaas/vald/blob/master/example/helm/values-standalone-agent-ngt.yaml) to deploy standalone Vald Agent.
    Each component can be disabled by setting the value `false` to the `[component].enabled` field.
    This is useful for deploying standalone Vald Agent NGT pods.

    ```bash
    helm repo add vald https://vald.vdaas.org/charts
    helm install vald-agent-ngt vald/vald --values example/helm/values-standalone-agent-ngt.yaml
    ```

1. Verify

    ```bash
    kubectl get pods
    ```

    <details><summary>Example output</summary><br>
    If the deployment is successful, Vald Agent component should be running.

    ```bash
    NAME               READY   STATUS    RESTARTS   AGE
    vald-agent-ngt-0   1/1     Running   0          20m
    vald-agent-ngt-1   1/1     Running   0          20m
    vald-agent-ngt-2   1/1     Running   0          20m
    vald-agent-ngt-3   1/1     Running   0          20m
    ```

    </details>

## Run Example Code

1. Port Forward

>     At first, pord fowarding the vald-lb-geateway is required to make request from your local environment possible.

Reviewer 2: ```suggestion
    At first, port forwarding the vald-lb-gateway is required to make request from your local environment possible.
```

    ```bash
    kubectl port-forward service/vald-agent-ngt 8081:8081
    ```

1. Download dataset

    Download [fashion-mnist](https://github.com/zalandoresearch/fashion-mnist) that is used as a dataset for indexing and search query.

    ```bash
    # move to working directory
    cd example/client/agent

    # download fashion-mnist testing dataset
    wget http://ann-benchmarks.com/fashion-mnist-784-euclidean.hdf5
    ```

1. Run Example

    We use [`example/client/agent/main.go`](https://github.com/vdaas/vald/blob/master/example/client/agent/main.go) to run the example.<br>
    This example will insert and index 400 vectors into the Vald from the fashion-mnist dataset via gRPC.

>     And then after waiting for indexing, it will request for searching hte nearest vector 10 times.

Reviewer: ```suggestion
    And then after waiting for indexing, it will request for searching the nearest vector 10 times.
```

                <details><summary>example code</summary><br>

                ```go
                const (
                    insertCount = 400
                    testCount = 20
                )
                ```

                </details>

            - The variables for configuration.
                <details><summary>example code</summary><br>

                ```go
                const (
                    datasetPath         string
                    grpcServerAddr      string
                    indexingWaitSeconds uint
                )
                ```

                </details>

        - Recognition parameters.
            <details><summary>example code</summary><br>

            ```go
            func init() {
                flag.StringVar(&datasetPath, "path", "fashion-mnist-784-euclidean.hdf5", "set dataset path")
                flag.StringVar(&grpcServerAddr, "addr", "127.0.0.1:8081", "set gRPC server address")
                flag.UintVar(&indexingWaitSeconds, "wait", 60, "set indexing wait seconds")
                flag.Parse()
            }
            ```

            </details>

    1. load

>         - Loading from fashion-mnist dataset and set id for each vector that is loaded. This step will return the training dataset, test dataset, and ids list of ids when loading is completed with success.
>             <details><summary>example code</summary><br>

Reviewer: ```suggestion
        - Loading from the fashion-mnist dataset and set id for each vector that is loaded. This step will return the training dataset, test dataset, and ids list of ids when loading is completed with success.
            <details><summary>example code</summary><br>
```

Author: I think it is not required too
